### PR TITLE
Api Client 2

### PR DESCRIPTION
**Pull request by jia614**
# Description

Apr-22-2022 01-21-12
Apr-22-2022 01-22-24

Created the `api-client` package for backend & search API interaction. 

Moved DTOs, search api functions, & our api functions to `api-client`.
Update all files to import from `@graduate/common` for DTO types.

Added dependency of `class-transformer`, which changes the `res.data` objects received from the backend to be instances of the appropriate response classes found in `@graduate/common/dtos`

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [X] This requires a run of `yarn install`

# How Has This Been Tested?

Please describe how you tested this PR (both manually and with tests)
Provide instructions so we can reproduce. 

- To test our API:
  - Run `yarn dev:v2` to launch frontend-v2 and api-v2
  - There are buttons on the frontend which tests various non-dev API calls to backend-v2 and the SearchAPI
  - Ensure that there are no existing users in local db before running tests as the user created may share the same uuid

# Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code where needed 
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Note:
- Please ignore all frontend v1 file changes!

### CODE REVIEW

**Comment by bugsalexander** [comment]
[Api client](https://trello.com/c/aUOEvB6R/426-api-client)

**Review by zefwang — APPROVED** [review]
Ligma got to me

File: packages/api-client/index.ts
Comment (bugsalexander) [inline packages/api-client/index.ts:170]:
```suggestion
    if (courseData?.class?.latestOccurrence) {
```

File: packages/api-client/index.ts
Comment (bugsalexander) [inline packages/api-client/index.ts:223]: move to top

File: packages/api-client/package.json
Comment (bugsalexander) [inline packages/api-client/package.json:21]: can remove this

File: packages/api-v2/package.json
Comment (bugsalexander) [inline packages/api-v2/package.json:24]: we should probably switch all the workspace imports to using `workspace:*`

File: packages/api-v2/src/auth/auth.controller.ts
Comment (bugsalexander) [inline packages/api-v2/src/auth/auth.controller.ts:22]: thoughts on changing the name back to student?

File: packages/api-v2/src/plan/plan.controller.ts
Comment (bugsalexander) [inline packages/api-v2/src/plan/plan.controller.ts:38]: same for this one

File: packages/api-v2/src/student/student.controller.ts
Comment (bugsalexander) [inline packages/api-v2/src/student/student.controller.ts:71]: we should write tests for backend asap (not this PR tho). @justineluo is there a spot in the timeline for implementing cypress?

File: packages/common/package.json
Comment (bugsalexander) [inline packages/common/package.json:6]: you can add something akin to sandbox/graduate team or something for author

File: packages/frontend-v2/pages/index.tsx
Comment (bugsalexander) [inline packages/frontend-v2/pages/index.tsx:5]: the two api-client imports can be merged

File: packages/frontend-v2/pages/index.tsx
Comment (bugsalexander) [inline packages/frontend-v2/pages/index.tsx:64]: i dont remember exactly how to implement auth,  but it shouldn't be part of the response body. i think it's either in the set-cookie header, or something else. cc @aryanshah701

File: packages/frontend/src/components/Sidebar/RequirementSection.tsx
Comment (bugsalexander) [inline packages/frontend/src/components/Sidebar/RequirementSection.tsx:11]: there are an awful lot of import and formatting changes in here– in the future, try to avoid changing files that you aren't touching functionally

File: packages/api-client/index.ts
Comment (BrandonLim8890) [reply packages/api-client/index.ts:48]: We pass the JWT into the header because the current `api-v2` doesn't support JWT via cookies. I talked to @aryanshah701 about this and he mentioned that for testing purposes, we should align the api-client to the current api-v2. Then once we make the ticket to switch the backend to utilise cookies, we will also change this accordingly

File: packages/api-v2/.eslintrc.js
Comment (BrandonLim8890) [reply packages/api-v2/.eslintrc.js:3]: I was also working with Aryan on this portion. Importing does work for the backend, but the way NestJS uses it's pipes doesn't allow us to import from common. If we were to import other objects/types for use throughout the backend it would work, it only won't work if we are using the imported type for NestJS pipes, which we do when we import the DTO's.

File: package.json
Comment (aryanshah701) [reply package.json:27]: yeap should be under dev!

File: packages/api-v2/.eslintrc.js
Comment (aryanshah701) [reply packages/api-v2/.eslintrc.js:3]: unfortunately building our DTOs using babel breaks class-validator validation pipes in the backend :( 

We're dealing with a ton of babel issues with building common so perhaps moving towards delegating the build step to the apps(frontend/backend) might be the move but I didn't want any of this to block them so we shall fix this in a separate ticket. Just made one https://trello.com/c/jHKm85F0/449-fix-dto-build-in-common

File: packages/api-v2/package.json
Comment (aryanshah701) [reply packages/api-v2/package.json:24]: ahh yes!

File: packages/api-v2/src/auth/auth.controller.ts
Comment (aryanshah701) [reply packages/api-v2/src/auth/auth.controller.ts:22]: hmmm... GetStudentResponse makes more sense to me since it decouples the response from the resource, but i'm ok with either

File: packages/api-v2/src/student/student.controller.ts
Comment (aryanshah701) [reply packages/api-v2/src/student/student.controller.ts:71]: Ooh, we could have integration tests that hit the endpoints using supertest written pretty quick since Nest is already setup with supertest. I normally use cypress for e2e testing from the frontend but if cypress has a ton of benefits then we can go with that

File: packages/frontend-v2/pages/index.tsx
Comment (aryanshah701) [reply packages/frontend-v2/pages/index.tsx:63]: since currently the backend is setup with access token authentication and not cookies, when logging in the token is returned in the body. 

my backend skills aren't great lmaoo but from all the research i did access token auth flow normally returns the token on signup in the response body, the client has full autonomy on how it wants to store it, and then the client can send it in the Authorization header as a bearer token on each request.

File: packages/api-client/index.ts
Comment (jia614) [reply packages/api-client/index.ts:140]: oops im sorry ;-; i gave misleading info-- did not know what a graphql client was apparently kek

File: .gitignore
Comment (aryanshah701) [inline .gitignore:96]: we need a .gitignore clean up lmao

File: packages/api-client/index.ts
Comment (bugsalexander) [reply packages/api-client/index.ts:140]: LOL it is okay <3

File: packages/api-v2/.eslintrc.js
Comment (bugsalexander) [reply packages/api-v2/.eslintrc.js:3]: sounds good!

File: packages/api-v2/src/student/student.controller.ts
Comment (bugsalexander) [reply packages/api-v2/src/student/student.controller.ts:71]: supertest would be good too! mostly anything, so we have some sort of a spec for the api

File: packages/api-client/index.ts
Comment (aryanshah701) [inline packages/api-client/index.ts:214]: let's remove the env var here cause the api is on the same domain as the frontend? and also cause it doesn't work haha

File: packages/frontend-v2/pages/index.tsx
Comment (bugsalexander) [reply packages/frontend-v2/pages/index.tsx:63]: i actually wasn't sure, so i did some research. but it seems like the way to go is via cookies 
- https://dev.to/gkoniaris/how-to-securely-store-jwt-tokens-51cf
- (the sample code != what they are saying, but the article seems to say cookies) https://javascript.plainenglish.io/the-ultimate-way-to-store-authentication-tokens-in-javascript-41900756c1ae
- https://stackoverflow.com/questions/51332747/where-to-store-auth-token-frontend-and-how-to-put-it-in-http-headers-of-multip

tldr: there are 3 main ways to store auth token: localstorage, sessionstorage, and cookie. however the first two are bad idea because token can be stolen by CSRF or XSS, so the preferred way is via cookies with some headers to make it secure.

brandon mentioned earlier we can introduce cookies later, though, which sounds good to me

File: packages/api-client/index.ts
Comment (bugsalexander) [inline packages/api-client/index.ts:46]: very nice!

**Review by bugsalexander — APPROVED** [review]

**Comment by bugsalexander** [comment]
approval = good to merge when changes are fixed

File: packages/frontend-v2/pages/index.tsx
Comment (aryanshah701) [inline packages/frontend-v2/pages/index.tsx:20]: do we need to uppercase every course when hitting search's api? if so we should move this to the fetchCourse function itslef

File: packages/frontend/config-overrides.js
Comment (aryanshah701) [inline packages/frontend/config-overrides.js:10]: do we not need this anymore??

Comment (aryanshah701) [reply packages/frontend/config-overrides.js:10]: if we don't know whether we need it or not/are confused, no point in doing research let's add it back and not touch the old frontend lmaooo

File: packages/frontend/src/index.tsx
Comment (aryanshah701) [inline packages/frontend/src/index.tsx:10]: nani

**Review by aryanshah701 — APPROVED** [review]
looks gooood to me!!! would like to see all of @bugsalexander comments addressed + CI passing before we merge though

File: packages/frontend/config-overrides.js
Comment (bugsalexander) [reply packages/frontend/config-overrides.js]: we don't need it because we're now using Create react app v5

File: packages/frontend/src/index.tsx
Comment (bugsalexander) [reply packages/frontend/src/index.tsx:10]: yea sorry i added this – typescript was complaining

Comment (bugsalexander) [reply packages/frontend/src/index.tsx:10]: its cuz of stricter types in react 18

File: packages/frontend/config-overrides.js
Comment (bugsalexander) [inline packages/frontend/config-overrides.js:1]: this can be deleted, we don't use react-app-rewire anymore

Comment (bugsalexander) [reply packages/frontend/config-overrides.js:1]: see my comment on aryans comment ^

**Pull request merged**
